File: ruby/sieve/sieve.rb
class Sieve

  def initialize(max)
    @range = 2..max
  end

  def primes
    primes = []
    composites = []

    @range.each do |curr|
      next if composites.include?(curr)
      primes << curr
      composites = composites | multiples(curr)
    end
    primes
  end

  def multiples(num)
    @range.select { |multiple| multiple % num == 0 }
  end

end

File: python/sieve/sieve.py
def sieve(limit):
    primes = []
    unmarked = list(range(2, limit + 1))
    while len(unmarked) > 0:
        prime = unmarked.pop(0)
        for multiplier in range(2, limit // prime + 1):
            if prime * multiplier in unmarked:
                unmarked.remove(prime * multiplier)
        primes.append(prime)
    return primes
